File: parserio/i386/Scan.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

function Scan(var buf; skip,size: Longint; str: String): Longint;

var
  result: Longint;

begin
  asm
        lea     edi,[str]
        mov     esi,edi
        xor     eax,eax
        lodsb
        stosb
        xor     ecx,ecx
        mov     ecx,eax
        xor     ebx,ebx
        mov     ebx,eax
        jecxz    @@9
@@1:    lodsb
        cmp     al,'a'
        jb      @@2
        cmp     al,'z'
        ja      @@2
        sub     al,20h
@@2:    stosb
        loop    @@1
        sub     edi,ebx
        mov     esi,[buf]
        add     esi,skip
        mov     ecx,size
        sub     ecx,skip
        jecxz   @@8
        cld
        sub     ecx,ebx
        jb      @@8
        inc     ecx
@@4:    mov     ah,[edi]
        and     ah,NOT 20h
@@5:    lodsb
        and     al,NOT 20h
        cmp     al,ah
        loopne  @@5
        jne     @@8
        dec     esi
        mov     edx,ecx
        mov     ecx,ebx
@@6:    repe    cmpsb
        je      @@10
        mov     al,[esi-1]
        cmp     al,'a'
        jb      @@7
        cmp     al,'z'
        ja      @@7
        sub     al,20h
@@7:    cmp     al,[edi-1]
        je      @@6
        sub     ecx,ebx
        add     esi,ecx
        add     edi,ecx
        inc     esi
        mov     ecx,edx
        jne     @@4
@@8:    xor     eax,eax
        jmp     @@11
@@9:    mov     eax,1
        jmp     @@11
@@10:   sub     esi,ebx
        mov     eax,esi
        sub     eax,dword ptr [buf]
        inc     eax
@@11:   dec     eax
        mov     result,eax
  end;
  Scan := result;
end;

File: parserio/Scan.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

int32_t Scan (const void *buf, ssize_t skip, ssize_t size,
              const String *str) {

  String _str[255+1];
  ssize_t len = Length (str), count = len;
  unsigned char *D = GetStr (_str);

  if (!len)
    return 0;

  CopyString (_str, str, 255);

  // Uppercase string
  do {
    unsigned char c = *D;

    if ((c >= 'a') && (c <= 'z'))
      c -= 'a' - 'A';

    *D = c;
    D++;
    count--;
  } while (count);

  count = size - skip - len + 1;
  if (count) {
    const unsigned char *S = (unsigned char *)buf + skip;
    D = GetStr (_str);

    do {
      ssize_t old_count;
      unsigned char a = *D & ~0x20, b;
      bool f;

      do {
        b = *S & ~0x20;
        S++;
        count--;
        f = a != b;
      } while (f && count);

      if (f)
        return -1;

      S--;
      old_count = count;
      count = len;

      do {
        f = false;
        while (count && !f) {
          f = *S == *D;
          S++;
          D++;
          count--;
        }

        if (f)
          return S - ((unsigned char *)buf + len);

        a = *(S - 1);
        if ((a >= 'a') && (a <= 'z'))
          a -= 'a' - 'A';
      } while (a == *(D - 1));

      S += count - len;
      D += count - len;
      count = old_count;
      S++;
    } while (S);
  }

  return -1;
}
